blackduck/Authentication.py

```diff
@@ -89,91 +89,76 @@ def authenticate(self):
-                self.http_exception_handler(
-                    response=response,
-                    name="authenticate"
-                )
-
-            content = response.json()
-            self.cookie = response.headers.get('Set-Cookie', 'COOKIE_UNSET')
-            self.auth_token = self.cookie[cookie.index('=')+1:cookie.index(';')]
-            self.valid_until = datetime.utcnow() + timedelta(milliseconds=int(content.get('expiresInMilliseconds', 900000)))
-
-        # Do not handle exceptions - just just more details as to possible causes
-        # Thus we do not catch a JsonDecodeError here even though it may occur
-        # - no futher details to give.  
-        except requests.exceptions.ConnectTimeout as connect_timeout:
-            logger.critical(f"could not establish a connection within {self.timeout}s, this may be indicative of proxy misconfiguration")
-            raise connect_timeout
-        except requests.exceptions.ReadTimeout as read_timeout:
-            logger.critical(f"slow or unstable connection, consider increasing timeout (currently set to {self.timeout}s)")
-            raise read_timeout
-        else:
-            logger.info(f"success: auth granted until {self.valid_until} UTC")
\ No newline at end of file
+            logger.warning("ssl verification disabled, connection insecure. do NOT use verify=False in production!")
+
+        credentials = {'j_username': self.username, 'j_password': self.password}
+        response = self.session.post(
+            url="/j_spring_security_check",
+            data=credentials,
+            auth=NoAuth()  # temporarily strip authentication to avoid infinite recursion
+        )
+
+        if response.status_code == 204:  # No Content
+            try:
+                cookie = response.headers['Set-Cookie']
+                self.auth_token = cookie[cookie.index('=')+1:cookie.index(';')]
+                self.valid_until = datetime.utcnow() + timedelta(minutes=120)  # token is good for 2 hours
+                logger.info(f"success: auth granted until {self.valid_until} UTC")
+                return
+            except (KeyError, ValueError):
+                logging.exception("HTTP response status code 204 but unable to obtain bearer token")
+                # fall through
+
+        if response.status_code == 401:
+            logging.error("HTTP response status code = 401 (Unauthorized)")
+            try:
+                logging.error(response.json()['errorMessage'])
+            except (json.JSONDecodeError, KeyError):
+                logging.exception("unable to extract error message")
+                logging.error("HTTP response headers: %s", response.headers)
+                logging.error("HTTP response text: %s", response.text)
+            raise RuntimeError("Unauthorized username/password", response)
+
+        # all unhandled responses fall through to here
+        logging.error("Unhandled HTTP response")
```

Comment:
The rest of this could probably be handled by http_exception_handler - I think it would be better to modify that to provide these details.

Reply:
Is the http_exception_handler mainly to provide a standard way for the library to report on the most common status_codes? What would be the way for someone to override it? Did you envision client.http_handler = my_function(resp, name)?

Reply:
Also note that for authentication only, I want to dump the full response headers and body if I get something other than the most common 200, 204 (token), or 401 paths. This may be unwarranted in the exception_handler for general API use.

Reply:
> Is the http_exception_handler mainly to provide a standard way for the library to report on the most common status_codes? What would be the way for someone to override it? Did you envision client.http_handler = my_function(resp, name)?
Either that or another `http_handler = http_handler or Blackduck.Utils.http_exception_handler` type of thing in `Client.__init__(..)`.